Source/Core/Core/HLE/HLE_VarArgs.h

```diff
@@ -0,0 +1,144 @@
+#pragma once
+
+#include "Common/Align.h"
+#include "Common/CommonTypes.h"
+
+#include "Core/HW/Memmap.h"
+#include "Core/PowerPC/PowerPC.h"
+
+#include <type_traits>
+
+namespace HLE
+{
+// See System V ABI (SVR4) for more details
+//  -> 3-18 Parameter Passing
+//  -> 3-21 Variable Argument Lists
+//
+// Source:
+// http://refspecs.linux-foundation.org/elf/elfspec_ppc.pdf
+class VAList
+{
+public:
+  explicit VAList(u32 stack, u32 gpr = 3, u32 fpr = 1, u32 gpr_max = 10, u32 fpr_max = 8)
+      : m_gpr(gpr), m_fpr(fpr), m_gpr_max(gpr_max), m_fpr_max(fpr_max), m_stack(stack)
+  {
+  }
+  ~VAList() = default;
+  // SFINAE
+  template <typename T>
+  struct IsArgPointer
+      : std::integral_constant<bool, std::is_union<T>::value || std::is_class<T>::value>
+  {
+  };
+  template <typename T>
+  struct IsWord : std::integral_constant<bool, std::is_pointer<T>::value ||
+                                                   (std::is_integral<T>::value && sizeof(T) <= 4)>
+  {
+  };
+  template <typename T>
+  struct IsDoubleWord : std::integral_constant<bool, std::is_integral<T>::value && sizeof(T) == 8>
+  {
+  };
+  template <typename T>
+  struct IsArgReal : std::integral_constant<bool, std::is_floating_point<T>::value>
+  {
+  };
+
+  // 0 - arg_ARGPOINTER
+  template <typename T, typename std::enable_if_t<IsArgPointer<T>::value>* = nullptr>
+  T GetArg()
+  {
+    std::aligned_storage_t<sizeof(T), alignof(T)> buffer;
+    u8* ptr = static_cast<u8*>(&buffer);
+    u32 addr = GetArg<u32>();
+
+    for (size_t i = 0; i < sizeof(T); i += 1, addr += 1)
+    {
+      ptr[i] = PowerPC::HostRead_U8(addr);
+    }
+
+    return *reinterpret_cast<T*>(ptr);
```

Comment:
Yeah, I know I was reading the standard and brainstorming to find a workaround. Unfortunately, I didn't find any "standard" way of doing it in C++. Each time there was a possible aliasing issue (not always detected by the compiler) that could allow an optimization to return (i.e. copy) the object before copying information into it. Using an union isn't doable (due to C++ active member restriction and) because the original point was to prevent the call of the default constructor in case the target type doesn't have one.

I fell back to a naive implementation that forces the target struct/union to have a default constructor. This case from the ABI isn't used by printf but might be useful for other HLE purposes.